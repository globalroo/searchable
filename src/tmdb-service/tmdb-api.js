import {
	getDiscoverMovieEndpoint,
	getMovieByIdEndpoint,
	getMovieCastByIdEndpoint,
	getMovieGenresEndpoint,
	getPersonByIdEndpoint,
	getPersonCreditsByIdEndpoint,
	getPopularMovieEndpoint,
	getRecommendedMoviesByIdEndpoint,
	getSearchMovieEndpoint,
	getSearchMultiEndpoint,
	getTrendingPersonByWeekEndpoint,
	getTrendingMoviesByDay,
	getTrendingMoviesByWeek,
	getTrendingTvByWeekEndpoint,
	getTvByIdEndpoint
} from "./tmdb-config";

export const MEDIA_TYPE = {
	TV: "tv",
	MOVIE: "movie",
	PERSON: "person"
};

export const defaultFailResponse = {
	page: 0,
	results: [],
	total_pages: 0,
	total_results: 0
};

const handleResponse = ({ response, payload }) => ({
	response,
	payload
});

/**
 * Naive safe response with capability to handle basic error state.
 * @async
 * @function safeFetchJson
 * @param {string} url - the endpoint to fetch
 * @param {object} options - fetch request options object default = {},
 * @param {object} failResponse - an expected empty response from tmdb default = defaultFailResponse
 * @return {object} { payload, response } - "payload" - holds JSON from the response or failResponse. "response" - holds the actual response from the server to handle error conditions
 */

export const safeFetchJson = async ({ url, options = {}, failResponse = defaultFailResponse }) => {
	let response, payload;
	try {
		response = await window.fetch(url, options);
		payload = await response.json();
	} catch (error) {
		response = { ok: false, statusText: `${error}` };
		payload = failResponse;
	}
	return handleResponse({ response, payload });
};

// API helpers

export const getPopularMovies = async () =>
	await safeFetchJson({
		url: getPopularMovieEndpoint()
	});

export const getSearchMovieTitleResults = async term =>
	await safeFetchJson({
		url: getSearchMovieEndpoint(term)
	});

export const getSearchMultiResults = async term =>
	await safeFetchJson({
		url: getSearchMultiEndpoint(term)
	});

export const getMovieById = async id =>
	await safeFetchJson({
		url: getMovieByIdEndpoint(id),
		failResponse: {}
	});

export const getRecommendedMoviesById = async id =>
	await safeFetchJson({
		url: getRecommendedMoviesByIdEndpoint(id)
	});

export const getMovieCastById = async id =>
	await safeFetchJson({
		url: getMovieCastByIdEndpoint(id)
	});

export const getPersonById = async id =>
	await safeFetchJson({
		url: getPersonByIdEndpoint(id),
		failResponse: {}
	});

export const getTvById = async id =>
	await safeFetchJson({
		url: getTvByIdEndpoint(id),
		failResponse: {}
	});

export const getDailyTrendingMovies = async () =>
	await safeFetchJson({
		url: getTrendingMoviesByDay()
	});

export const getWeeklyTrendingMovies = async () =>
	await safeFetchJson({
		url: getTrendingMoviesByWeek()
	});

export const getWeeklyTrendingPeople = async () =>
	await safeFetchJson({
		url: getTrendingPersonByWeekEndpoint()
	});

export const getWeeklyTrendingTv = async () =>
	await safeFetchJson({
		url: getTrendingTvByWeekEndpoint()
	});

export const getPersonCreditsById = async id =>
	await safeFetchJson({
		url: getPersonCreditsByIdEndpoint(id)
	});

export const getMovieGenres = async () =>
	await safeFetchJson({
		url: getMovieGenresEndpoint()
	});

export const getDiscoverMoviesByGenreId = async (id, query) => {
	const applyFilter = query
		? query
		: "&sort_by=popularity.desc&include_adult=false&vote_average.gte=6";
	return await safeFetchJson({
		url: `${getDiscoverMovieEndpoint()}${applyFilter}&with_genres=${id}`
	});
};
