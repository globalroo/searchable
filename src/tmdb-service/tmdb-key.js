let API_KEY = ""; //TODO: handle API key

export const getKey = () => {
	return API_KEY;
};
