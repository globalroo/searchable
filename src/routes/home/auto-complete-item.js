import React from "react";
import { MenuItem } from "@material-ui/core";
import { Movie, Tv, Person } from "@material-ui/icons/";
import { MEDIA_TYPE } from "src/tmdb-service/tmdb-api.js";
import { Typography } from "@material-ui/core";

const getAutoCompleteItemIcon = media_type => {
	switch (media_type) {
		case MEDIA_TYPE.MOVIE:
			return <Movie />;
		case MEDIA_TYPE.TV:
			return <Tv />;
		case MEDIA_TYPE.PERSON:
			return <Person />;
		default:
			return null;
	}
};

export const AutoCompleteItem = ({ item, index, highlighted, selected, ...props }) => (
	<MenuItem {...props} selected={highlighted} component="div">
		{getAutoCompleteItemIcon(item.media_type)}
		<Typography>{`${item.title || item.name}`}</Typography>
	</MenuItem>
);
