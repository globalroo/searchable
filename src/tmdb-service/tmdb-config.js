import { getKey } from "./tmdb-key";

export const BASE_URL = "https://api.themoviedb.org/3";
export const BASE_IMG_PATH = "http://image.tmdb.org/t/p";

const API_KEY = getKey();

// Obtain image sizes here
// https://api.themoviedb.org/3/configuration?api_key=

export const API = {
	COMBINED_CREDITS: "/combined_credits",
	CREDITS: "/credits",
	DAY: "/day",
	DISCOVER_MOVIE: "/discover/movie",
	GENRES: "/genre/movie/list",
	MOVIE: "/movie",
	PERSON: "/person",
	POPULAR_MOVIES: "/movie/popular",
	RECOMMENDATIONS: "/recommendations",
	SEARCH_MOVIES: "/search/movie",
	SEARCH_MULTI: "/search/multi",
	SMALL_IMG_WIDTH: "/w300",
	LARGE_IMG_WIDTH: "/w1280",
	TRENDING: "/trending",
	TV: "/tv",
	WEEK: "/week"
};

export const getAuthorisedEndpoint = endpoint => `${BASE_URL}${endpoint}?api_key=${API_KEY}`;

export const getSmallPosterImage = imagePath =>
	`${BASE_IMG_PATH}${API.SMALL_IMG_WIDTH}/${imagePath}`;

export const getLargeImage = imagePath => `${BASE_IMG_PATH}${API.LARGE_IMG_WIDTH}/${imagePath}`;

export const getMovieByIdEndpoint = id =>
	getAuthorisedEndpoint(`${API.MOVIE}/${id}`) + "&append_to_response=recommendations,credits";

export const getMovieCastByIdEndpoint = id =>
	getAuthorisedEndpoint(`${API.MOVIE}/${id}${API.CREDITS}`);

export const getPopularMovieEndpoint = () => getAuthorisedEndpoint(API.POPULAR_MOVIES);

export const getRecommendedMoviesByIdEndpoint = id =>
	getAuthorisedEndpoint(`${API.MOVIE}/${id}${API.RECOMMENDATIONS}`);

export const getSearchMovieEndpoint = searchTerm =>
	`${getAuthorisedEndpoint(API.SEARCH_MOVIES)}&query=${searchTerm}`;

export const getSearchMultiEndpoint = searchTerm =>
	`${getAuthorisedEndpoint(API.SEARCH_MULTI)}&query=${searchTerm}`;

export const getTrendingMoviesByDay = () =>
	getAuthorisedEndpoint(`${API.TRENDING}${API.MOVIE}${API.DAY}`);

export const getTrendingMoviesByWeek = () =>
	getAuthorisedEndpoint(`${API.TRENDING}${API.MOVIE}${API.WEEK}`);

export const getTrendingPersonByWeekEndpoint = () =>
	getAuthorisedEndpoint(`${API.TRENDING}${API.PERSON}${API.WEEK}`);

export const getTrendingTvByWeekEndpoint = () =>
	getAuthorisedEndpoint(`${API.TRENDING}${API.TV}${API.WEEK}`);

export const getPersonByIdEndpoint = id =>
	getAuthorisedEndpoint(`${API.PERSON}/${id}`) + "&append_to_response=images";

export const getPersonCreditsByIdEndpoint = id =>
	getAuthorisedEndpoint(`${API.PERSON}/${id}${API.COMBINED_CREDITS}`);

export const getTvByIdEndpoint = id =>
	getAuthorisedEndpoint(`${API.TV}/${id}`) + "&append_to_response=recommendations,credits";

export const getMovieGenresEndpoint = () => getAuthorisedEndpoint(`${API.GENRES}`);
export const getDiscoverMovieEndpoint = () => getAuthorisedEndpoint(`${API.DISCOVER_MOVIE}`);
