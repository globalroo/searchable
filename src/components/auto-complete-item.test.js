import React from "react";
import { render, cleanup } from "react-testing-library";
import { MEDIA_TYPE } from "../tmdb-service/tmdb-api";
import { AutoCompleteItem } from "./auto-complete-item";

import "jest-dom/extend-expect";
const mockMovieItem = {
	id: 0,
	title: "movieItem",
	media_type: MEDIA_TYPE.MOVIE
};

const mockTvItem = {
	id: 1,
	title: "tvItem",
	media_type: MEDIA_TYPE.TV
};

const mockPersonItem = {
	id: 2,
	name: "personItem",
	media_type: MEDIA_TYPE.PERSON
};

describe("Auto complete item tests", () => {
	afterEach(cleanup);
	it("Renders an auto complete component with a movie item", () => {
		const mockData = {
			item: mockMovieItem,
			selected: false,
			highlighted: false
		};
		const { container } = render(<AutoCompleteItem {...mockData} />);
		expect(container.firstChild).toMatchSnapshot();
	});
	it("Renders an auto complete component with a tv item", () => {
		const mockData = {
			item: mockTvItem,
			selected: false,
			highlighted: false
		};
		const { container } = render(<AutoCompleteItem {...mockData} />);
		expect(container.firstChild).toMatchSnapshot();
	});
	it("Renders an auto complete component with a person item", () => {
		const mockData = {
			item: mockPersonItem,
			selected: false,
			highlighted: false
		};
		const { container } = render(<AutoCompleteItem {...mockData} />);
		expect(container.firstChild).toMatchSnapshot();
	});
	it("Renders an auto complete component with an unhandled item", () => {
		const restoreConsole = mockConsole();
		expect(() => render(<AutoCompleteItem item={{ media_type: "unknown" }} />)).toThrowError();
		restoreConsole();
	});
	it("Passes through additional props", () => {
		const mockData = {
			item: mockPersonItem,
			selected: false,
			highlighted: false,
			className: "additional-prop"
		};
		const { container } = render(<AutoCompleteItem {...mockData} />);
		expect(container.firstChild).toHaveClass("additional-prop");
	});
});
