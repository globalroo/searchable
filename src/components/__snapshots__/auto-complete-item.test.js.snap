// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auto complete item tests Renders an auto complete component with a movie item 1`] = `
<div
  class="MuiButtonBase-root-16 MuiListItem-root-4 MuiListItem-default-7 MuiListItem-gutters-12 MuiListItem-button-13 MuiMenuItem-root-1 MuiMenuItem-gutters-2"
  data-testid="0"
  role="menuitem"
  tabindex="-1"
>
  <svg
    aria-hidden="true"
    class="MuiSvgIcon-root-19"
    focusable="false"
    role="presentation"
    viewBox="0 0 24 24"
  >
    <path
      d="M18 4l2 4h-3l-2-4h-2l2 4h-3l-2-4H8l2 4H7L5 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V4h-4z"
    />
    <path
      d="M0 0h24v24H0z"
      fill="none"
    />
  </svg>
   
  <p
    class="MuiTypography-root-28 MuiTypography-body1-37"
  >
    movieItem
  </p>
  <span
    class="MuiTouchRipple-root-64"
  />
</div>
`;

exports[`Auto complete item tests Renders an auto complete component with a person item 1`] = `
<div
  class="MuiButtonBase-root-156 MuiListItem-root-144 MuiListItem-default-147 MuiListItem-gutters-152 MuiListItem-button-153 MuiMenuItem-root-141 MuiMenuItem-gutters-142"
  data-testid="2"
  role="menuitem"
  tabindex="-1"
>
  <svg
    aria-hidden="true"
    class="MuiSvgIcon-root-159"
    focusable="false"
    role="presentation"
    viewBox="0 0 24 24"
  >
    <path
      d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
    />
    <path
      d="M0 0h24v24H0z"
      fill="none"
    />
  </svg>
   
  <p
    class="MuiTypography-root-168 MuiTypography-body1-177"
  >
    personItem
  </p>
  <span
    class="MuiTouchRipple-root-204"
  />
</div>
`;

exports[`Auto complete item tests Renders an auto complete component with a tv item 1`] = `
<div
  class="MuiButtonBase-root-86 MuiListItem-root-74 MuiListItem-default-77 MuiListItem-gutters-82 MuiListItem-button-83 MuiMenuItem-root-71 MuiMenuItem-gutters-72"
  data-testid="1"
  role="menuitem"
  tabindex="-1"
>
  <svg
    aria-hidden="true"
    class="MuiSvgIcon-root-89"
    focusable="false"
    role="presentation"
    viewBox="0 0 24 24"
  >
    <path
      d="M0 0h24v24H0z"
      fill="none"
    />
    <path
      d="M21 3H3c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h5v2h8v-2h5c1.1 0 1.99-.9 1.99-2L23 5c0-1.1-.9-2-2-2zm0 14H3V5h18v12z"
    />
  </svg>
   
  <p
    class="MuiTypography-root-98 MuiTypography-body1-107"
  >
    tvItem
  </p>
  <span
    class="MuiTouchRipple-root-134"
  />
</div>
`;
